tools/codegen/dest/register_dispatch_key.py

```diff
@@ -56,14 +56,29 @@ def gen_create_out_helper(backend_index: BackendIndex) -> List[str]:
         dispatch = str(backend_index.dispatch_key).lower()
         empty_impl = f"at::detail::empty_{dispatch}"
         empty_strided_impl = f"at::detail::empty_strided_{dispatch}"
+        runtime_quantized_check = ""
     elif backend_index.dispatch_key == DispatchKey.CompositeExplicitAutograd:
         empty_impl = "at::empty"
         empty_strided_impl = "at::empty_strided"
+        runtime_quantized_check = """\
+  if (isQIntType(typeMetaToScalarType(options.dtype()))) {{
+    // This dynamic check is mainly to improve error messages for structured operators
+    // that do not have a quantized kernel implementation.
+    // If the tensor is quantized then we can't call at::empty(), because quantized
+    // kernels have their own dedicated operator, and the empty call will fail.
+    // Instead, we'll create an undefined tensor and call the out= operator.
+    // If the out-of-place op doesn't have a quantized kernel (which is why we ended up here),
+    // then the out= variant of the op typically won't either, so we'll error out with
+    // a nicer error message, complaining that the out= operator doesn't have a quantized implementation.
+    return at::Tensor();
```

Comment:
so by doing this, does it guarantee that composite out= operator (e.g. structured native_batch_norm) fails, just by virtue of noticing one of its inputs is undefined?  

i'm not following how this results in an error message that implicates missing an implementation for quantized backend.  it seems more like it would just fail due to an undefined input tensor.

but I have a feeling you're one step ahead of me and this is probably fine :)

Reply:
It actually doesn't guarantee that the composite out= operator fails.

I'm basically assuming that if `{op}` doesn't have a quantized kernel, then `{op}.out` won't have one either. If you call `{op}` with quantized tensors, then because of the codegen'd kernel here, we'll funnel you to the `{op}.out` operator (and pass it an undefined output tensor). As long as `{op}.out` doesn't have a quantized implementation, then we'll fail in the *dispatcher* by complaining that `{op}.out` doesn't have a quantized implementation.

Technically, if there were an op that the quantization team happened to write a kernel for the out= variant but not the out-of-place variant, then with this PR it would have to handle undefined tensors properly if you wanted to be able to call the out-of-place variant. That's strictly better than today though, where calling the out-of-place variant will just unconditionally fail (because we try to call `at::empty()`, which will never work for quantized).

Long term, a better fix would be to add real support for quantized backends to structured kernels (there's actually an open issue for it: https://github.com/pytorch/pytorch/issues/52680). That is definitely more involved though, so this PR will just avoid making things worse until then.

Let me know if

Reply:
ahh ok that makes sense.  ok, then i agree its strictly better. thanks!